File: src/main/java/architecture/goldenboughs_lib/common/payload/toc/PlayerDamagePayload.java
package architecture.goldenboughs_lib.common.payload.toc;

import architecture.goldenboughs_lib.api.LcDamageType;
import architecture.goldenboughs_lib.client.gui.hudlayers.screenfilter.LcDamageScreenFilterLayer;
import architecture.goldenboughs_lib.common.payload.api.ToClientPayload;
import architecture.goldenboughs_lib.core.GoldenBoughsLib;
import architecture.goldenboughs_lib.util.PayloadUtil;
import io.netty.buffer.ByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

public record PlayerDamagePayload(
	Optional<@Nullable LcDamageType> lcDamageType,
	float damage
) implements ToClientPayload {
	public static final Type<PlayerDamagePayload> TYPE = new Type<>(GoldenBoughsLib.modRl("player_damage_payload"));

	public static final StreamCodec<ByteBuf, PlayerDamagePayload> STREAM_CODEC = StreamCodec.composite(
		ByteBufCodecs.optional(LcDamageType.STREAM_CODEC), PlayerDamagePayload::lcDamageType,
		ByteBufCodecs.FLOAT, PlayerDamagePayload::damage,
		PlayerDamagePayload::new
	);

	public PlayerDamagePayload(@Nullable LcDamageType lcDamageType, float damage) {
		this(Optional.ofNullable(lcDamageType), damage);
	}

	public static void send(ServerPlayer player, LcDamageType lcDamageType, float newDamage) {
		PayloadUtil.sendToPlayerClient(player, new PlayerDamagePayload(lcDamageType, newDamage));
	}

	@Override
	public @NotNull CustomPacketPayload.Type<? extends CustomPacketPayload> type() {
		return TYPE;
	}

	@Override
	public void work(Player player) {
		LcDamageScreenFilterLayer.INSTANCE.addFilter(lcDamageType().orElse(null));
	}
}

File: src/main/kotlin/architecture/goldenboughs_lib/common/payload/toc/PlayerDamagePayload.kt
package architecture.goldenboughs_lib.common.payload.toc

import architecture.goldenboughs_lib.api.LcDamageType
import architecture.goldenboughs_lib.api.payload.ToClientPayload
import architecture.goldenboughs_lib.client.gui.hudlayers.screenfilter.LcDamageScreenFilterLayer
import architecture.goldenboughs_lib.core.Lib
import architecture.goldenboughs_lib.util.PayloadUtil.sendToPlayerClient
import io.netty.buffer.ByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import net.neoforged.neoforge.network.handling.IPayloadContext
import java.util.*

@JvmRecord
data class PlayerDamagePayload(
	@JvmField val lcDamageType: Optional<LcDamageType>,
	@JvmField val damage: Float
) : ToClientPayload {
	constructor(lcDamageType: LcDamageType, damage: Float) : this(
		Optional.ofNullable(lcDamageType),
		damage
	)

	override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> {
		return TYPE
	}

	override fun work(context: IPayloadContext, player: Player) {
		LcDamageScreenFilterLayer.INSTANCE.addFilter(this.lcDamageType.orElse(null))
	}

	companion object {
		@JvmField
		val TYPE: CustomPacketPayload.Type<PlayerDamagePayload> =
			CustomPacketPayload.Type(Lib.modRl("player_damage_payload"))

		@JvmField
		val STREAM_CODEC: StreamCodec<ByteBuf, PlayerDamagePayload> =
			StreamCodec.composite(
				ByteBufCodecs.optional(LcDamageType.STREAM_CODEC),
				PlayerDamagePayload::lcDamageType,
				ByteBufCodecs.FLOAT,
				PlayerDamagePayload::damage,
				::PlayerDamagePayload
			)

		@JvmStatic
		fun send(player: ServerPlayer, lcDamageType: LcDamageType, newDamage: Float) {
			sendToPlayerClient(player, PlayerDamagePayload(lcDamageType, newDamage))
		}
	}
}

